### PR TITLE
DP-1384: start pactbroker as non-root on k8s

### PR DESCRIPTION
**Pull request by patvong**
#### Description
We have `pactbroker` running on the `tools.infra.use1` cluster currently as root using the image https://hub.docker.com/r/dius/pact-broker/dockerfile.
The Docker image is based off `phusion/passenger-ruby24:1.0.0` which runs `CMD ["/sbin/my_init"]` that requires root user for many things.
To workaround many issues (requiring root), I had to modify the Dockerfile to be able to run as non-root with user `app`:
* add the `app` user to the `root` group
* modified the ownerships of variaous files and directories 
* disabled syslog and cron
* expose 8080 port instead of port 80 from nginx

Please review @AppDirect/ad-need4speed 

I tested this docker image with the k8s manifest configuration as follow:
### pactbroker.yml
```
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: null
  labels:
    io.kompose.service: pact-broker
  name: pact-broker
  namespace: default
spec:
  ports:
  - name: "8080"
    port: 8080
    targetPort: 8080
  selector:
    io.kompose.service: pact-broker
  type: LoadBalancer

---

apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  creationTimestamp: null
  labels:
    io.kompose.service: pact-broker
  name: pact-broker
  namespace: default
spec:
  replicas: 1
  strategy: {}
  template:
    metadata:
      creationTimestamp: null
      labels:
        io.kompose.service: pact-broker
    spec:
      containers:
      - env:
        - name: PACT_BROKER_DATABASE_USERNAME
          value: postgres
        - name: PACT_BROKER_DATABASE_PASSWORD
          value: postgres
        - name: PACT_BROKER_DATABASE_HOST
          value: postgres
        - name: PACT_BROKER_DATABASE_NAME
          value: postgres
        - name: PACT_BROKER_PUBLIC_HEARTBEAT
          value: "true"
        # load docker image locally
        imagePullPolicy: Never
        image: my-pact-broker:latest
        name: pact-broker
        securityContext:
          runAsNonRoot: true
          runAsUser: 9999
        ports:
        - containerPort: 8080
        resources: {}
        livenessProbe:
          tcpSocket:
            port: 8080
          initialDelaySeconds: 30
          timeoutSeconds: 10
        readinessProbe:
          tcpSocket:
            port: 8080
          initialDelaySeconds: 30
          timeoutSeconds: 10
      restartPolicy: Always
```

### postgres.yml
```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  creationTimestamp: null
  labels:
    io.kompose.service: postgres-claim0
  name: postgres-claim0
  namespace: default
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 100Mi

---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  creationTimestamp: null
  labels:
    io.kompose.service: postgres
  name: postgres
  namespace: default
spec:
  replicas: 1
  strategy:
    type: Recreate
  template:
    metadata:
      creationTimestamp: null
      labels:
        io.kompose.service: postgres
    spec:
      containers:
      - env:
        - name: POSTGRES_PASSWORD
          value: postgres
        - name: POSTGRES_USER
          value: postgres
        image: postgres:9.6-alpine
        name: postgres
        resources: {}
        volumeMounts:
        - mountPath: /var/lib/postgresql/data
          name: postgres-claim0
      restartPolicy: Always
      volumes:
      - name: postgres-claim0
        persistentVolumeClaim:
          claimName: postgres-claim0

---
apiVersion: v1
kind: Service
metadata:
  creationTimestamp: null
  labels:
    io.kompose.service: postgres
  name: postgres
  namespace: default
spec:
  clusterIP: None
  ports:
  - name: headless
    port: 55555
    targetPort: 0
  selector:
    io.kompose.service: postgres
```

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -11,13 +11,32 @@ RUN bash -lc 'rvm --default use ruby-2.4.5'
...
     gem install --no-document --minimal-deps bundler && \
     bundle install --deployment --without='development test' && \
     rm -rf vendor/bundle/ruby/2.4.0/cache/ /usr/local/rvm/rubies/ruby-2.4.4/lib/ruby/gems/2.4.0/cache
 
-EXPOSE 80
+EXPOSE 8080
+
+# Adding 'app' user to group 'root'
+RUN set -ex && \
+    adduser app root
```
Comment (duboisf) [inline Dockerfile]: What's the `root` group?

File: Dockerfile [hunk 2]
```diff
@@ -11,13 +11,32 @@ RUN bash -lc 'rvm --default use ruby-2.4.5'
...
+
+# Disabling syslog that needs to be run as root
+RUN rm -rf /etc/my_init.d/10_syslog-ng.init
+# Disabling cron that needs to be run as root
+RUN touch /etc/service/cron/down && chmod +x /etc/service/cron/down
+
 CMD ["/sbin/my_init"]
+
+# uid is 9999
+USER app
```
Comment (duboisf) [inline Dockerfile]: You can do `USER 9999`, that way you don't need to specify the non-root UID in the kubernetes manifest

Comment (duboisf) [reply Dockerfile]: (`runAsUser 9999`)

File: Dockerfile (hunk 1 above)
Comment (patvong) [reply Dockerfile]: It's the group (with id = 0) associated to the `root` user.
This is inspired by Openshift guidelines: https://docs.openshift.com/enterprise/3.2/creating_images/guidelines.html#openshift-enterprise-specific-guidelines
```
For an image to support running as an arbitrary user, directories and files that may be written to by processes in the image should be owned by the root group and be read/writable by that group. Files to be executed should also have group execute permissions.
```

**Pull request closed without merging**